### ISSUE TITLE
how to test the model?

### ISSUE THREAD

**Issue by apeterswu** [miyosuda/async_deep_reinforce#9]
Hi, I am new to Tensorflow and interested in running this project. 
But I don't see test descriptions in your readme wiki. Could you please give a description how to test the model? 
Thanks a lot.

**Comment by miyosuda** [comment]
First you need to install ALE by following the instruction written in 
https://github.com/miyosuda/async_deep_reinforce#how-to-build

If you prefer OpenAI gym instead of ALE, you can use "gym" branch.

Learned model data is not included in repository because of size, so you need to run learning process to see the learning result.

To run learning, 

    $python a3c.py

After running for hours, please stop learning with hitting "Ctrl-C" key in console to stop learning. Then learned data is saved.

After that you can see the game result with.

    $python a3c_disp.py

**Comment by apeterswu** [comment]
Thanks for your quick response, Miyosuda. I was confused before. Currently I am using your project to do some other problems. Would you mind give me an email address that I can contact with you, appreciated if you may help me for some others? Thanks very much.

**Comment by apeterswu** [comment]
Thanks for your quick response. I have one other problem, I noticed you set gradient norm clip to be 40, could you tell me how to choose this number? Thanks in advance.

**Comment by miyosuda** [comment]
I followed the advice given by @aravindsrinivas in issues thread.
https://github.com/miyosuda/async_deep_reinforce/issues/1#issuecomment-214360003

> The maximum allowed gradient norm was 40.

**Comment by apeterswu** [comment]
Hi, Miyosuda, I have one question while testing. When you are doing test, choose action is also done by randomly select one action but according by their distribution, but why not just choose the largest probability action???

**Comment by lforg37** [comment]
Hello,

Following the probability distribution allows to find non deterministic policies which could be better than deterministic ones when the state perception isn't complete and differents states leads to the same perception but the optimal action isn't necessarily the same in thoose states.

I think that it might not be useful in some games like breakout in which I suppose that 4 game frames could be sufficient to sum up the internal game state but it could be useful in some other games in which for example you only see à small part of the screen at a time. 

However, I'm not sure it's still useful when the agent has some kind of memory (like lstm cells).

**Comment by apeterswu** [comment]
@lforg37 Thanks for your reply. Can I understand like this, whether deterministic is decided by the task, in some task, maybe deterministic policy is better, better while in the game like breakout or pong, maybe stochastic policy is better.

**Comment by lforg37** [comment]
@apeterswu I think it depends more on how much information you get from the environnent than on intrinsics characteristics of the tasks. If you could have all the informations from the environnent internal state, you could allways have an optimal deterministic policy. But I don't know how you define a task so if it includes the given information from the environnent I agree with you.

And for lstm, they modify the perceived state so even if two environnement states give the same perceptions, the lstms (or whatever internal memory the agent uses) car be used to distinguish one from the other. So I think it could eliminate the need of using a stochastic policy.

**Comment by apeterswu** [comment]
Hi, @miyosuda , I got one more problem. In your original code, save model only happens when we use "ctrl+c". I want to save the intermediate model, so I modified the "a3c.py" in "train_function":
```python
while True:
    if global_t > log_k * 500000 and parallel_index == 0:
      print global_t
      if not os.path.exists(CHECKPOINT_DIR):
        os.mkdir(CHECKPOINT_DIR)
      saver.save(sess, CHECKPOINT_DIR + '/' + 'checkpoint', global_step = global_t)
      log_k = log_k + 1

    if stop_requested:
      break
    if global_t > MAX_TIME_STEP:
      break

    diff_global_t = training_thread.process(sess, global_t, summary_writer,
                                            summary_op, score_input)
    global_t += diff_global_t
```
I defined "log_k" to be global just like "global_t", but I found one problem is that, when it successfully saved some model, e.g. 50w, 100w, 200w.., later, if saved 400w, it would cover former saved model/delete former saved model, which means 20w,100w...saved model are gone, the directory would only have 400w model... I am very confused about this problem, and have no idea why it happens. Could you please help? Thanks.

**Comment by miyosuda** [comment]
> https://www.tensorflow.org/versions/r0.10/api_docs/python/state_ops.html#Saver

> max_to_keep indicates the maximum number of recent checkpoint files to keep. As new files are created, older files are deleted. If None or 0, all checkpoint files are kept. Defaults to 5 (that is, the 5 most recent checkpoint files are kept.)

How about applying 0 for "max_to_keep"?

**Comment by apeterswu** [comment]
@miyosuda Thanks a lot. I think it will help.

**Comment by sahiliitm** [comment]
@apeterswu 
The reason for drawing from probability distribution over discrete action space is two-fold:
1. During Training this helps in exploration. Suppose you think that action 1 is the best and current policy has assigned it probability of 0.8. It might happen due to stochasticity of policy that you pick action 4 which had probability 0.05. Now if this action gave a reaaaaaally big reward, then it would make sense to increase the probability of this action, right? This would never have been possible if you kept picking the max probability action again and again.

2. During testing time, whether drawing from distribution differs from picking max probability action depends on the kind of policy you have converged to:

Case 1: The policy is very **soft** meaning that no one action has very high (near 1) probability. In this case it is true that selecting max probability action would yield to better performance. However, if you do not select actions with randomness, you have no idea how well the agent does in certain parts of the state-space. The agent could learn to play game in a way such that it always scores a very high score but ONLY for a single trajectory of (observations, actions). To guard against this, the best way to evaluate stochastic **soft** policies is to do some kind of an epsilon-greedy evaluation. High score under this scheme would point to the fact that the agent can handle some level of stochasticity and performs well in many parts of the state space. I think the ideal way to evaluate is to range epsilon from 0 to 1 and plot a curve of average score versus epsilon. This gives a very good idea of how the agent can handle being in _difficult_ parts of the state space.

Case 2: If the policy you have converged to is hard (say that for any state there exists a single action with probability of picking that action > 0.9999) then it probably won't matter which evaluation procedure is followed.

**Comment by apeterswu** [comment]
@miyosuda @lforg37 , currently I am using this framework to do another problem(the optimal policy will be a deterministic policy), but replace the CNN to be FF, and use lstm, but found no improvement after trained more than 2000w steps, what kind of elements do you think would be the reasons? Is it because this is a deterministic problem?

**Comment by apeterswu** [comment]
@sahiliitm, Thanks for your response. I agree with your idea during training, I am just curious about the performance of this two ways. Just like my question on last comment, I do know I have a problem which the best policy would be a deterministic policy, do you think soft training and soft test would be a good solution?

**Comment by lforg37** [comment]
@apeterswu  Maybe the hyperparameters (learning rate, ...) are not optimised for your problem ? Or your initial weights are not good ? Without any more informations on the problem it is difficult to have a clue on what should have been wrong.

@sahiliitm 

> Case 1: The policy is very soft meaning that no one action has very high (near 1) probability. In this case it is true that selecting max probability action would yield to better performance.

I don't agree with that statement. Let's imagine you have the following problem (taken from David Silver reinforcement learning course) : an agent is on a discrete map. 
The agent can only perceive the following attribute from it's place : SW, EW,NW, WW which are booleans which are true if there is a wall on south, east, north, west.

Let's take the following map : 
```
 _____ _____ _____ _____ _____
|                             |
|  1     2     3     4     5  |
|      _____       _____      |
|     |     |     |     |     |
|  6  |     |  7  |     |  8  |
|_____|     |_____|     |_____|
```
Let's say there is a trap in the places 6 and 8 and there is a treasure in the place 7.
Your reward is -10 when you get into places 6 and 8  and 10 when you get into place 7. Places 7, 6 and 8 are final states.

If you have a deterministic policy, once your learning is done you will always choose the same action for the same perceived state. The problem is that for places 2 and 4 the perceived state is the same (i.e. SW and NW are true, EW and WW are false) so you would have to choose the same action when you are in 2 or in 4. 

Or this is clearly a bad idea because if you choose for example to always go to the left, when you get in place 2 depending on your best action for 1 you either loop infinitely in (1,2,1,2...) or you do (1,2,6). If you choose to always go right the problem is the same when you get in place 4.

If you have a stochastic policy, you could learn that when you perceive (SW && NW && !EW && !WW) you have to go left in 50% of the time and right the other 50%. This way, you could do a small loop (1,2,1,2...) but you know that eventually you will reach 3. The optimal stochastic policy given the perception is thus

```
 _____ _____ _____ _____ _____
|                             |
|  >    < >    v    < >    <  |
|      _____       _____      |
|     |     |     |     |     |
|  X  |     |  X  |     |  X  |
|_____|     |_____|     |_____|
```

**Comment by sahiliitm** [comment]
@lforg37 
Under the assumption that the observations could be aggregated to get a MDP-like state, my statement is true :)
The case you mention is that of a POMDP. Typically in atari domain with 4 concatenated states we get a reasonably good approximation to a MDP-state.

In your case, the state can be the concatenation of last 4 observations.

Besides my statement was meant to be an "emperical truth" in the sense that this is what I have observed in my experiments with Atari. Certainly, there could exist synthetic examples where it is not true. But from my experiments, sampling from A3C-Learnt policies (for my modification of A3C) led to average scores of about 7000 on sea quest whereas not sampling but picking max probability actions (90% of times) lead to scores of about 18000.

**Comment by lforg37** [comment]
@sahiliitm I'm sorry in this case. I think you were talking in the general case.

**Comment by sahiliitm** [comment]
@lforg37  thanks for pointing out that case :) I hadn't considered it.

**Comment by apeterswu** [comment]
@sahiliitm owo, I think your result is good. Could you explain more about your action chosen during training and testing? That would be very helpful. Thanks a lot.

@miyosuda , One small question about LOCAL_T_MAX, you write the comment: "#repeat step size", it is repeat one action size? not the forward step (which means n-step)?

**Comment by apeterswu** [comment]
@miyosuda , if I want to add one more LSTM layer for policy output, and one more another LSTM layer for value output, how to modify the network, could you help just write an example, thanks a lot.
